tensorflow_quantum/python/optimizers/spsa_minimizer.py

```diff
@@ -0,0 +1,275 @@
+        Numpy array (if static value is obtainable), else `tf.Tensor`.
+    """
+    static_x = tf.get_static_value(x)
+    if static_x is not None:
+        return static_x
+    return x
+
+
+SPSAOptimizerResults = collections.namedtuple(
+    'SPSAOptimizerResults',
+    [
+        'converged',
+        # Scalar boolean tensor indicating whether the minimum
+        # was found within tolerance.
+        'num_iterations',
+        # The number of iterations of the SPSA update.
+        'num_objective_evaluations',
+        # The total number of objective
+        # evaluations performed.
+        'position',
+        # A tensor containing the last argument value found
+        # during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        # A tensor containing the value of the objective from
+        # previous iteration
+        'objective_value_previous_iteration',
+        # Save the evaluated value of the objective function
+        # from the previous iteration
+        'objective_value',
+        # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'tolerance',
+        # Define the stop criteria. Iteration will stop when the
+        # objective value difference between two iterations is
+        # smaller than tolerance
+        'a',
+        # Specifies the learning rate
+        'alpha',
+        # Specifies scaling of the learning rate
+        'c',
+        # Specifies the size of the perturbations
+        'gamma',
+        # Specifies scaling of the size of the perturbations
+        'blocking',
+        # If true, then the optimizer will only accept updates that improve
+        # the objective function.
+        'allowed_increase'
+        # Specifies maximum allowable increase in objective function
+        # (only applies if blocking is true).
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function,
+                       a, alpha, c, gamma, blocking, allowed_increase):
+    """Create SPSAOptimizerResults with initial state of search."""
+    init_args = {
+        "converged": tf.Variable(False),
+        "num_iterations": tf.Variable(0),
+        "num_objective_evaluations": tf.Variable(0),
+        "position": tf.Variable(initial_position),
+        "objective_value": tf.Variable(0.),
+        "objective_value_previous_iteration": tf.Variable(0.),
+        "tolerance": tolerance,
+        "a": tf.Variable(a),
+        "alpha": tf.Variable(alpha),
+        "c": tf.Variable(c),
+        "gamma": tf.Variable(gamma),
+        "blocking": tf.Variable(blocking),
+        "allowed_increase": tf.Variable(allowed_increase)
+    }
+    return SPSAOptimizerResults(**init_args)
+
+
+def minimize(expectation_value_function,
+             initial_position,
+             tolerance=1e-7,
```

Comment:
Might want to lower this to 1e-4 or 1e-5. Float32's aren't reliable to this many decimals.

Reply:
Changed to 1e-5